applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp

```diff
@@ -0,0 +1,754 @@
+    KRATOS_TRY
+
+    const GeometryType& Geom = this->GetGeometry();
+    const unsigned int N_DOF = TNumNodes * (TDim + 1);
+
+    if ( rValues.size() != N_DOF )
+        rValues.resize( N_DOF, false );
+
+    if ( TDim > 2 )
+    {
+        unsigned int index = 0;
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( DISPLACEMENT_X, Step );
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( DISPLACEMENT_Y, Step );
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( DISPLACEMENT_Z, Step );
+            rValues[index++] = 0.0;
+        }
+    }
+    else
+    {
+        unsigned int index = 0;
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( DISPLACEMENT_X, Step );
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( DISPLACEMENT_Y, Step );
+            rValues[index++] = 0.0;
+        }
+    }
+
+    KRATOS_CATCH( "" )
+}
+
+//----------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwBaseElement<TDim,TNumNodes>::
+    GetFirstDerivativesVector( Vector& rValues, int Step ) const
+{
+    KRATOS_TRY
+
+    const GeometryType& Geom = this->GetGeometry();
+    const unsigned int N_DOF = TNumNodes * (TDim + 1);
+
+    if ( rValues.size() != N_DOF )
+        rValues.resize( N_DOF, false );
+
+
+    if ( TDim > 2 )
+    {
+        unsigned int index = 0;
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( VELOCITY_X, Step );
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( VELOCITY_Y, Step );
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( VELOCITY_Z, Step );
+            rValues[index++] = 0.0;
+        }
+    }
+    else
+    {
+        unsigned int index = 0;
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( VELOCITY_X, Step );
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( VELOCITY_Y, Step );
+            rValues[index++] = 0.0;
+        }
+    }
+
+    KRATOS_CATCH( "" )
+}
+
+//----------------------------------------------------------------------------------------
+
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwBaseElement<TDim,TNumNodes>::GetSecondDerivativesVector( Vector& rValues, int Step ) const
+{
+    KRATOS_TRY
+
+    const GeometryType& Geom = this->GetGeometry();
+    const unsigned int N_DOF = TNumNodes * (TDim + 1);
+
+    if ( rValues.size() != N_DOF )
+        rValues.resize( N_DOF, false );
+
+    unsigned int index = 0;
+
+    if ( TDim > 2 )
+    {
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            rValues[index++] = Geom[i].FastGetSolutionStepValue( ACCELERATION_X, Step );
```

Comment:
here you are accessing 3 times the database.

it would be faster if you did

~~~cpp
const auto& acc  = Geom[i].FastGetSolutionStepValue( ACCELERATION, Step );
rValues[index++] = acc[0];
rValues[index++] = acc[1]
...
~~~